MF-170 - Add manager API unit tests

anovakovic01 opened the pull request:
This pull request resolves #170.

**Summary:**
- Add unit test for every manager API endpoint
- Update swagger documentation
- Fix connection request validation
- Fix EOF error handling
- Add invalid content type error handling
- Update README with codecov badge

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

drasko reviewed (dismissed):
LGTM

anovakovic01 committed:
Fix connection request validation and EOF error handling

Fix validation of connection and client id in connection request.
Fix EOF error handling by returning HTTP status Bad Request.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Add manager API tests and update swagger

Implement unit tests for every manager API endpoint. Update client
and connection mock implementations by switching to UUIDs. Update
swagger file with correct status codes.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+	test(svc, ts.URL, client)
+}
+
+func sendRequest(client *http.Client, method string, url string, contentType string, token string, body io.Reader) (*http.Response, error) {
```
Please rename to `makeRequest`, and wrap all of the parameters in a structure called `testRequest`.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+			{"register new user", string(validBody), http.StatusCreated},
+			{"try to register existing user", string(validBody), http.StatusConflict},
+			{"try to register user with invalid email address", string(invalidBody), http.StatusBadRequest},
+			{"try to register with invalid request format", "{", http.StatusBadRequest},
```
Empty request would be viable test case as well.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+			status int
+		}{
+			{"register new user", string(validBody), http.StatusCreated},
+			{"try to register existing user", string(validBody), http.StatusConflict},
```
Please drop "try to". This applies for all specs.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+	return manager.New(users, clients, channels, hasher, idp)
+}
+
+func runService(test func(manager.Service, string, *http.Client)) {
```
Instead of passing a test callback, rename this method to `newServer` and return its instance in the same way it is done at `newService`. You can obtain all of the data in the test itself.

Just to be clear, this is not a bad way to remove duplication, but I'd prefer that we are as explicit as possible in our test cases.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+			status int
+			res    string
+		}{
+			{"login with valid credentials", string(validCredentials), http.StatusCreated, string(token)},
```
These `string(...)` are hard to read. Don't "torture" the table, do the cast during test data preparation. This applies to each of the further occurrences.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+
+var (
+	validUser       = manager.User{"user@example.com", "password"}
+	invalidPassUser = manager.User{"user@example.com", "invalid_password"}
```
Since this is used only once, you can drop it from here and create an instance locally, based on the `validUser`.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+var (
+	validUser       = manager.User{"user@example.com", "password"}
+	invalidPassUser = manager.User{"user@example.com", "invalid_password"}
+	invalidUser     = manager.User{"userexample.com", "password"}
```
This "fellow" exists for the sole purpose of providing a false e-mail value. Define a constant that contains "bad" value instead (this can be reused for other purposes as well, see other tests as an example).

The only exception to the aforementioned statement is user registration. For that purpose you can create local instance based on the `validUser` but using the constant I mentioned as an e-mail.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+	invalidUser     = manager.User{"userexample.com", "password"}
+	nonexistentUser = manager.User{"non-existentuser@example.com", "pass"}
+	validClient     = manager.Client{Type: "app", Name: "test_app", Payload: "test_payload"}
+	invalidClient   = manager.Client{Type: "foo", Name: "invalid_client", Payload: "some_payload"}
```
Same remark as `invalidUser` and `invalidPassUser`.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+const contentType = "application/json; charset=utf-8"
+
+var (
+	validUser       = manager.User{"user@example.com", "password"}
```
Please drop "valid" prefix from all valid entities. This will improve readability together with other changes.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+		req.Header.Set("Authorization", token)
+	}
+	if contentType != "" {
+		req.Header.Set("Content-Type", contentType)
```
Add test cases that cover the missing content type. Since this might be the edge case, we should expect `415` error.

anovakovic01 replied:
If content type is missing, handlers ignore that fact and work normally, so there is no need for this test case.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+
+func TestUpdateClient(t *testing.T) {
+	runService(func(svc manager.Service, url string, client *http.Client) {
+		validUpdate, _ := json.Marshal(manager.Client{
```
Why do you create new "valid" instance. Reuse!

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+			Name:    "test_device",
+			Payload: "device_payload",
+		})
+		invalidUpdate, _ := json.Marshal(manager.Client{
```
Create invalid based on the existing valid. This will make the "error" more explicit.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+			Payload: "device_payload",
+		})
+		svc.Register(validUser)
+		clientID, _ := svc.AddClient(validUser.Email, validClient)
```
Please rename to `id`, it is sufficient.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+func TestViewClient(t *testing.T) {
+	runService(func(svc manager.Service, url string, client *http.Client) {
+		svc.Register(validUser)
+		clientID, _ := svc.AddClient(validUser.Email, validClient)
```
Please rename to `id`.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+		cases := []struct {
+			desc     string
+			req      string
+			clientID string
```
Please rename to `ID`.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+			status   int
+		}{
+			{"update existing client", string(validUpdate), clientID, validUser.Email, http.StatusOK},
+			{"try to update non-existent client", string(validUpdate), "123e4567-e89b-12d3-a456-000000000042", validUser.Email, http.StatusNotFound},
```
Wrong uuid can be made constant instead of repeating it.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+			status int
+			res    []manager.Client
+		}{
+			{"fetch list of clients", validUser.Email, http.StatusOK, clients},
```
Maybe add another case where we retrieve clients for user without any clients (you should create that user as well).

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+func TestRemoveClient(t *testing.T) {
+	runService(func(svc manager.Service, url string, client *http.Client) {
+		svc.Register(validUser)
+		clientID, _ := svc.AddClient(validUser.Email, validClient)
```
Rename to `id`.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+		}{
+			{"create new channel", string(validData), validUser.Email, http.StatusCreated},
+			{"try to create new channel with invalid token", string(validData), invalidUser.Email, http.StatusForbidden},
+			{"try to create new channel with invalid data format", "{", validUser.Email, http.StatusBadRequest},
```
Please add the empty JSON test.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+			{"add valid client", string(validClient), validUser.Email, http.StatusCreated},
+			{"try to add client with invalid data", string(invalidClient), validUser.Email, http.StatusBadRequest},
+			{"try to add client with invalid auth token", string(validClient), "invalid_token", http.StatusForbidden},
+			{"try to add client with invalid request format", "}", validUser.Email, http.StatusBadRequest},
```
Please add the empty JSON test.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+			{"try to update client with invalid id", string(validUpdate), "1", validUser.Email, http.StatusNotFound},
+			{"try to update client with invalid data", string(invalidUpdate), clientID, validUser.Email, http.StatusBadRequest},
+			{"try to update client with invalid user token", string(validUpdate), clientID, invalidUser.Email, http.StatusForbidden},
+			{"try to update client with invalid data format", "{", clientID, validUser.Email, http.StatusBadRequest},
```
Please add the empty JSON test.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+			{"delete existing client", clientID, validUser.Email, http.StatusNoContent},
+			{"try to delete non-existent client", "123e4567-e89b-12d3-a456-000000000042", validUser.Email, http.StatusNoContent},
+			{"try to delete client with invalid id", "1", validUser.Email, http.StatusNoContent},
+			{"try to delte client with invalid token", clientID, invalidUser.Email, http.StatusForbidden},
```
Fix typo.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+			"name": "updated_channel",
+		})
+		svc.Register(validUser)
+		chanID, _ := svc.CreateChannel(validUser.Email, validChannel)
```
Remarks given for client-related tests regarding asking to rename to `id` apply here as well.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+			{"try to update non-existing channel", string(update), "123e4567-e89b-12d3-a456-000000000042", validUser.Email, http.StatusNotFound},
+			{"try to update channel with invalid token", string(update), chanID, invalidUser.Email, http.StatusForbidden},
+			{"try to update channel with invalid id", string(update), "1", validUser.Email, http.StatusNotFound},
+			{"try to update channel with invalid data format", "}", chanID, validUser.Email, http.StatusBadRequest},
```
Please add the empty JSON test.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+		}{
+			{"connect existing client to existing channel", chanID, clientID, validUser.Email, http.StatusOK},
+			{"try to connect existing client to non-existent channel", "123e4567-e89b-12d3-a456-000000000042", clientID, validUser.Email, http.StatusNotFound},
+			{"try to connect client with invalid id to channel", chanID, "1", validUser.Email, http.StatusNotFound},
```
Add tests that will try to connect client from user A to the channel from user B and vice versa.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+			auth     string
+			status   int
+		}{
+			{"disconnect connected client from channel", chanID, clientID, validUser.Email, http.StatusNoContent},
```
Add tests that try to disconnect channel from user A from client from user B and vice versa. This is an upgrade to the invalid token test :)

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+	runService(func(svc manager.Service, url string, client *http.Client) {
+		svc.Register(validUser)
+		clientID, _ := svc.AddClient(validUser.Email, validClient)
+		unconnectedClientID, _ := svc.AddClient(validUser.Email, validClient)
```
Please rename to `notConnectedClientID`.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,535 @@
+			status    int
+			clientID  string
+		}{
+			{"validate token over existing channel id", chanID, clientID, http.StatusOK, clientID},
```
Change descriptions to use the following "template":  `check access to [...] given [...]`.

mijicd commented on manager/mocks/channels.go:
```diff
@@ -29,7 +28,7 @@ func (crm *channelRepositoryMock) Save(channel manager.Channel) (string, error)
 
 	crm.counter += 1
-	channel.ID = strconv.Itoa(crm.counter)
+	channel.ID = fmt.Sprintf("123e4567-e89b-12d3-a456-%012d", crm.counter)
```
Theoretically, this will start giving invalid UUIDs after 9 entities. 

Instead, let's be more explicit about the way we want to use this implementation. Predefine an array of n valid uuids, and cycle through it using the counter modulo n as an index. Yes, there will be duplicates, but this should be a stub anyhow, and later on we can improve its logic if it is necessary.

mijicd replied:
Please ignore this comment. I misunderstood your implementation.

mijicd commented on manager/mocks/clients.go:
```diff
@@ -29,7 +28,7 @@ func (crm *clientRepositoryMock) Id() string {
 
 	crm.counter += 1
-	return strconv.Itoa(crm.counter)
+	return fmt.Sprintf("123e4567-e89b-12d3-a456-%012d", crm.counter)
```
See the remark given for channels mock.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,718 @@
+	return httptest.NewServer(mux)
+}
+
+func makeRequest(client *http.Client, testReq testRequest) (*http.Response, error) {
```
Even better approach is to bind this method, and the `http.Client` to the `testRequest`. Its signature would be `testRequest.make()`.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,718 @@
+	ts := newServer(svc)
+	defer ts.Close()
+	client := ts.Client()
+	url := ts.URL
```
Drop this variable.

mijicd commented on manager/api/transport_test.go:
```diff
@@ -0,0 +1,718 @@
+	client := ts.Client()
+	url := ts.URL
+
+	body, _ := json.Marshal(user)
```
Please create helper `toJson(data interface{}) string`, and use it instead of creating "artificial" temporary names.

mijicd reviewed (changes requested):
Please correct these minor changes along with the explicitly given remark:
- add `Content-Type` check in code and documentation
- try to integrate with codecov.io

anovakovic01 committed:
Add content type check and update documentation

Add content type check in implementation and update documentation
accordingly. Refactor tests and add empty content type test cases.
Add code coverage badge to readme.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>
